_posts/2019-01-04-nextjs.md

```diff
@@ -0,0 +1,290 @@
+- projects were not updated when new versions of react.js were released
+
+The first successful solution was the Next.js library, which, has 3600 contributors and over 33K “stars” on Github. To evaluate the benefits of this library, consider what kind of functionality you need to provide for the operation of a universal web application:
+
+- Asynchronous data loading. In order for the results of asynchronous requests to be received prior to server rendering, Next.js implements a special type of the page component, which has an asynchronous static async getInitialProps ({req}) life cycle event.
+
+- Server-side rendering (SSR).
+
+- Passing server component state to client.
+
+- Creating a component on the side of a web browser and linking it to an HTML document.
+
+- Routing on the server and on the client should also be universal. That is, the same definition of routing should work for both server and client code.
+
+- Code splitting. For each page, only the necessary JavaScript code should be loaded, not the entire application. When moving to the next page, the missing code should be reloaded - without reloading the same modules, and without unnecessary modules.
+
+All these tasks are successfully solved by the Next.js library. This library is based on a very simple idea. It is proposed to introduce a new type of component - “page”, in which there is an async static method async getInitialProps ({req}). The “page” component is a normal React component. This type of component can be thought of as a new type in the series: “component”, “container”, “page”.
+
+## Sample Project
+
+Let's create a simple project with Next.js to demonstrate how it works. You will be surprised how little we need to do in order to start our actual application. Some complexity is hidden by the library and the rule convention over configuration is helping to keep code on our side really slim.
+
+Create a new directory for our sample project `mkdir sample-app && cd sample-app`. Generate an empty npm project here with `npm init -y`. Then install Next.js together with React dependencies: `npm i next react react-dom --save`.
+
+Now we just have to add a new npm run script to our `package.json`:
+
+```js
+  "scripts": {
+    "dev": "next"
+  }
+```
+
+Let's try: `npm run dev`. Unfortunately you'll see an error message in your terminal. But with  next.js, error messages are usually pretty good and tell you exactly what you have to do next:
+
+`Couldn't find a `pages` directory. Please create one under the project root`
+
+So go ahead, create a `pages` directory and rerun `npm run dev`. It will tell you that it is now running on port 3000 - so open localhost:3000 in your browser. It's rendering a 404 page however because we did not yet create any "page".
+
+<img src="/img/posts/2019-01-04-nextjs/404.png" alt="404 Not Found" class="centered" />
+
+Lets create an `index.js` file with the following content:
+
+`export default () => <div>Hello World</div>`
```

Comment:
Missing comma! ;) Commas save lives! Example: `Let's eat grandma.` vs `Let's eat, grandma.` ;)

(Here and in other examples also)

Reply:
I miss commas all the time, but in this case I think it really isn't necessary?

Reply:
Well, it is the world's most recognizable code example ... ;) Up to you.

Reply:
Aaah, I think now I get it. I really thought there is a comma missing here: `Lets create an `index.js` file with the following content`.

Regarding the `Hello World`, I also have an excuse for that. The german wikipedia page states that `Hello World!` without comma started becoming more common :D (I missed the `!` however.